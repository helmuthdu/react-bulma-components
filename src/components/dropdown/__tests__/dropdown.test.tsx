import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from 'jsdom';
import * as React from 'react';
import { Dropdown, DropdownDivider, DropdownItem } from '..';

declare const global: any;

describe('Dropdown component', () => {
  beforeEach(() => {
    global.window = new JSDOM('<body><div id="app-root"></div></body>').window;
  });

  it('should render', () => {
    const { asFragment } = render(
      <Dropdown value="value" onChange={() => {}}>
        <DropdownItem value="value">Item</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should add listener do document on mount', () => {
    global.window.document.addEventListener = jest.fn();
    render(
      <Dropdown value="value" onChange={() => {}}>
        <DropdownItem value="value">Item</DropdownItem>
      </Dropdown>
    );
    expect(window.document.addEventListener).toHaveBeenCalled();
  });

  it('should concat classname in props with classname', () => {
    const { asFragment } = render(
      <Dropdown value="value" className="other-class" onChange={() => {}}>
        <DropdownItem value="value">Item</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have custom inline styles', () => {
    const { asFragment } = render(
      <Dropdown value="value" style={{ width: 400 }} onChange={() => {}}>
        <DropdownItem value="value">Item</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have divider', () => {
    const { asFragment } = render(
      <Dropdown value="value" style={{ width: 400 }} onChange={() => {}}>
        <DropdownItem value="value">Item</DropdownItem>
        <DropdownDivider />
        <DropdownItem value="other">Other</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be right-aligned when using "right" prop', () => {
    const { asFragment } = render(
      <Dropdown right>
        <DropdownItem value="value">Item</DropdownItem>
        <DropdownItem value="other">Other</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should appear above the dropdown button', () => {
    const { asFragment } = render(
      <Dropdown up>
        <DropdownItem value="value">Item</DropdownItem>
        <DropdownItem value="other">Other</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should open the Dropdown', () => {
    const { container } = render(
      <Dropdown value="value" style={{ width: 400 }} onChange={() => {}}>
        <DropdownItem value="value">Item</DropdownItem>
        <DropdownDivider />
        <DropdownItem value="other">Other</DropdownItem>
      </Dropdown>
    );
    const dropdown = container.querySelector('.dropdown') as HTMLDivElement;
    const button = dropdown.querySelector('button') as HTMLButtonElement;
    expect(dropdown).not.toHaveClass('is-active');
    fireEvent.click(button);
    expect(dropdown).toHaveClass('is-active');
  });

  it('should select the new value', async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Dropdown hoverable value="" style={{ width: 400 }} onChange={onChange}>
        <DropdownItem value="value">Foo</DropdownItem>
      </Dropdown>
    );
    const dropdownItem = container.querySelector('.dropdown-item') as HTMLDivElement;
    const button = container.querySelector('button') as HTMLButtonElement;
    fireEvent.mouseOver(button);
    fireEvent.click(dropdownItem);
    expect(onChange).toHaveBeenCalledWith('value');
  });

  it('should select the new value and close the dropdown', async () => {
    const onChange = jest.fn();
    const { container } = render(
      <Dropdown value="" style={{ width: 400 }} onChange={onChange}>
        <DropdownItem value="value">Foo</DropdownItem>
      </Dropdown>
    );
    const dropdown = container.querySelector('.dropdown') as HTMLDivElement;
    const dropdownItem = container.querySelector('.dropdown-item') as HTMLDivElement;
    const button = container.querySelector('button') as HTMLButtonElement;
    fireEvent.click(button);
    fireEvent.click(dropdownItem);
    expect(onChange).toHaveBeenCalledWith('value');
    expect(dropdown).not.toHaveClass('is-active');
  });

  it('should show custom label passed to the label prop', () => {
    const { asFragment } = render(
      <Dropdown label="test label">
        <DropdownItem value="value">Item</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show the label of the first dropdown item when no custom label is passed', () => {
    const { asFragment } = render(
      <Dropdown>
        <DropdownItem value="value">Item</DropdownItem>
      </Dropdown>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show custom label when active valued is undefined/empty', () => {
    const { container } = render(
      <Dropdown label="test label" value="">
        <DropdownItem value="value">Item</DropdownItem>
      </Dropdown>
    );
    const button = container.querySelector('button') as HTMLButtonElement;
    expect(button).toHaveTextContent('test label');
  });

  it('should show the label of the dropdown item when value of it is the active value', () => {
    const { container } = render(
      <Dropdown label="test label" value="value">
        <DropdownItem value="value">Bar</DropdownItem>
      </Dropdown>
    );
    const button = container.querySelector('button') as HTMLButtonElement;
    expect(button).toHaveTextContent('Bar');
  });
});
