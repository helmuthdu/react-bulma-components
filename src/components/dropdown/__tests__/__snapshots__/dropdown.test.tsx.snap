// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should appear above the dropdown button 1`] = `
<DocumentFragment>
  <div
    class="dropdown is-up"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          Item
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item"
        >
          Item
        </a>
        <a
          class="dropdown-item"
        >
          Other
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown component should be right-aligned when using "right" prop 1`] = `
<DocumentFragment>
  <div
    class="dropdown is-right"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          Item
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item"
        >
          Item
        </a>
        <a
          class="dropdown-item"
        >
          Other
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown component should concat classname in props with classname 1`] = `
<DocumentFragment>
  <div
    class="dropdown other-class"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          Item
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item is-active"
        >
          Item
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown component should have custom inline styles 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
    style="width: 400px;"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          Item
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item is-active"
        >
          Item
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown component should have divider 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
    style="width: 400px;"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          Item
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item is-active"
        >
          Item
        </a>
        <hr
          class="dropdown-divider"
        />
        <a
          class="dropdown-item"
        >
          Other
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown component should render 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          Item
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item is-active"
        >
          Item
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown component should show custom label passed to the label prop 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          test label
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item"
        >
          Item
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Dropdown component should show the label of the first dropdown item when no custom label is passed 1`] = `
<DocumentFragment>
  <div
    class="dropdown"
  >
    <div
      role="presentation"
    >
      <button
        class="button"
        tabindex="0"
      >
        <span>
          Item
        </span>
        <span
          class="icon is-normal"
        >
          <i
            class="mdi mdi-chevron-down mdi-24px"
          />
        </span>
      </button>
    </div>
    <div
      class="dropdown-menu"
      role="menu"
    >
      <div
        class="dropdown-content"
      >
        <a
          class="dropdown-item"
        >
          Item
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
